### ISSUE TITLE
Incorrect swapchain format check for version 1.0

### ISSUE THREAD

**Issue by p2jason** [KhronosGroup/Vulkan-ValidationLayers#5185]
**Environment:**
 - OS: Windows 10
 - GPU: NVIDIA RTX 2060
 - SDK or header version if building from repo: 1.3.240
 - Options enabled (synchronization, best practices, etc.): Synchronization (altough, the issue happens with and without sync. validation)

**Describe the Issue**

The layer reports that the image format passed to `vkCreateSwapchainKHR` through `VkSwapchainCreateInfoKHR::imageFormat` is not supported, even though, the format is supported by the device. This happens for all formats returned by `vkGetPhysicalDeviceSurfaceFormatsKHR`.

<details>
<summary>Error message sent by the validation layer</summary>

```
Validation Error: [ VUID-VkSwapchainCreateInfoKHR-imageFormat-01778 ] Object 0: handle = 0x12228e24ba0, type = VK_OBJECT_TYPE_DEVICE; | MessageID = 0xc036022f | vkCreateSwapchainKHR(): pCreateInfo->imageFormat VK_FORMAT_B8G8R8A8_UNORM with tiling VK_IMAGE_TILING_OPTIMAL has no supported format features on this physical device. The Vulkan spec states: The implied image creation parameters of the swapchain must be supported as reported by vkGetPhysicalDeviceImageFormatProperties
```

</details>

**Expected behavior**

The layer shouldn't be reporting the format as not supported.

**Additional context**

After some investigating, it seems that the issue is caused by `CoreChecks::GetPDFormatProperties`. The function's return value (which is of type `VkFormatProperties3KHR`) doesn't get filled, causing `optimalTilingFeatures` to be zero and therefore triggering an error message in `CoreChecks::ValidateCreateSwapchain`.

In `CoreChecks::GetPDFormatProperties`, the `has_format_feature2` flag is set to true, so `fmt_props_2` and `fmt_props_3` are filled out by `DispatchGetPhysicalDeviceFormatProperties2`. However, it seems that for Vulkan version 1.0, this function doesn't follow `fmt_props_2`'s pNext chain, so `fmt_props_3` remains empty. Initally, I thought this might have to do with the fact that `VkFormatProperties3` doesn't become core until Vulkan 1.3 and I hadn't enabled the `VK_KHR_format_feature_flags2` extension, so I tried changing the Vulkan version. It turns out that `fmt_props_3` gets filled out properly for version 1.1, 1.2, and 1.3, so this is a problem exclusively on Vulkan 1.0.

**Comment by spencer-lunarg** [comment]
Yes, (**from a quick look now**) we are not handling `VK_KHR_format_feature_flags2` correctly in `GetPDFormatProperties()`... will get time tomorrow to look closer into this, easy fix, but need to make sure we have proper test coverage to prevent this from happing

**Comment by spencer-lunarg** [comment]
ok, so there is a nasty subtle issue here it seems, `vkGetPhysicalDeviceFormatProperties2` was added with `VK_KHR_get_physical_device_properties2` and promoted to core 1.0

But the code is 

```cpp
 if (has_format_feature2) {
        DispatchGetPhysicalDeviceFormatProperties2(physical_device, format, &fmt_props_2);
 }
```

because `VK_KHR_format_feature_flags2` is a special extension, if you read the spec it will never talk about "enabling" the extension, but "if it is supported"... if the driver exposes the `VK_KHR_format_feature_flags2` extension it is automatically enabled

This opens a case for a Vulkan 1.0 instance with a driver exposing `VK_KHR_format_feature_flags2` to not be able to use `GetPhysicalDeviceFormatProperties2` correctly

**Comment by spencer-lunarg** [comment]
actually looking at the spec, it says in `VK_KHR_format_feature_flags2`

> Requires [VK_KHR_get_physical_device_properties2](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#VK_KHR_get_physical_device_properties2) to be enabled for any device-level functionality

so, it should not be possible to hit this case, but still, the function seems to need to be fixed

**Comment by spencer-lunarg** [comment]
@p2jason I was not able to fully reproduce this, assuming you built the layers to tests this, can you try applying the patch in https://github.com/KhronosGroup/Vulkan-ValidationLayers/pull/5187 and seeing if that solves this for you on Vulkan 1.0

If not, then it is a driver bug and want to report/fix it and add CTS coverage

**Comment by p2jason** [comment]
I applied the patch, but it seems I have run into another issue. Apparently, the Vulkan version that is passed to the validation layers is Vulkan 1.1 (4198400) instead of what I specified, which is Vulkan 1.0 (4194304). This is according to the information gathered from `CreateInstance` in *chassis.cpp*. My guess is that the driver is changing this for some reason. For reference, here is the stack trace I got from a breakpoint placed in `CreateInstance`:

> ... (function where vkCreateInstance is called) ...
> vulkan-1.dll!00007ffab20ff4e7()
> vulkan-1.dll!00007ffab20e3d83()
> nvoglv64.dll!00007ffa853ec592()
> graphics-hook64.dll!00007ffad7aca115()
> VkLayer_khronos_validation.dll!vulkan_layer_chassis::CreateInstance(const VkInstanceCreateInfo* pCreateInfo, const VkAllocationCallbacks* pAllocator, VkInstance_T** pInstance)

I'm mentioning this because it means that the check you added (`api_version == VK_API_VERSION_1_0`) doesn't pass and therefore, `DispatchGetPhysicalDeviceFormatProperties2KHR` never gets called. However, even after commenting out the if statement so that `DispatchGetPhysicalDeviceFormatProperties2KHR` always gets called, it still didn't work. I get the same problem as before where the pNext chain isn't followed.

Also, I turned on warnings and I'm now getting these messages:

> vkGetPhysicalDeviceProperties2KHR: Emulation found unrecognized structure type in pProperties->pNext - this struct will be ignored
> vkGetPhysicalDeviceFormatProperties2: Emulation found unrecognized structure type in pFormatProperties->pNext - this struct will be ignored

I am getting these messages for both the pre-patch and post-patch versions of `CoreChecks::GetPDFormatProperties`, which is odd because I have not enabled device emulation, at least not manually. 

I am starting to think this might be a driver problem.

**Comment by p2jason** [comment]
Ok, so, after some further inverstigation, it seems I have figured out the issue. I have OBS installed and, apparently, it adds its own layer to intercept graphics calls (its that `graphics-hook64.dll` from the stack trace in the previous comment). I disabled OBS's layer and everything is working fine now. The api version is what I specified, I'm not getting the warning messages about emulation, and more importantly, I am no longer getting an error when callig `vkCreateSwapchainKHR`.

If this checks out with you, I'll go raise the issue with the OBS folks.

**Comment by spencer-lunarg** [comment]
Ok, first thanks again for spending more time on this.

Another question is was it set as

Option 1: `app` -> `OBS` -> `Validation` -> `Driver`
Option 2: `app` -> `Validation` -> `OBS` ->  `Driver`

and if you swap it dose that matter?

without having the OBS source to scan through, my only guess is OBS is just not having some generated code to handle the `GetPhysicalDeviceFormatProperties2` correctly for the new `VK_KHR_format_feature_flags2` system adding a `VkFormatProperties3` on everyone 🙃

**Comment by spencer-lunarg** [comment]
closing for now, if you want to reopen with more info to work off with feel free

**Comment by p2jason** [comment]
Sorry for the late response. It took a while for me to investigate things. I couldn't figure out how to change the order of the layers, however... At some point, I cloned the Vulkan loader to try and get it to load the layers in a different order, and everything worked ok (I was on the latest commit and I hadn't made any changes).

I went back to version `1.3.236` (which is the version of the SDK I have installed) and everything still worked. Then, I went back to version `1.3.224` (which is the version of the loader installed on my system) and I got the errors again! So, I downloaded the latest version of the runtime from LunarG and everything is fine now. I guess this was a loader issue and not an OBS issue?

Anyway, I'm not going to reopen the issue because it's seems like the problem has already been fixed.
